<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <meta name="author" content="Peter Haverty">
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>API - GenomicRanges.jl</title>
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="../css/highlight.css">
  <link href="../assets/Documenter.css" rel="stylesheet">
  
  <script>
    // Current page data
    var mkdocs_page_name = "API";
    var mkdocs_page_input_path = "api.md";
    var mkdocs_page_url = "/api/";
  </script>
  
  <script src="../js/jquery-2.1.1.min.js"></script>
  <script src="../js/modernizr-2.8.3.min.js"></script>
  <script type="text/javascript" src="../js/highlight.pack.js"></script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> GenomicRanges.jl</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
	<ul class="current">
	  
          
            <li class="toctree-l1">
		
    <a class="" href="..">Home</a>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">News</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../NEWS/">Changelog</a>
                </li>
                <li class="">
                    
    <a class="" href="../TODO/">TODO</a>
                </li>
    </ul>
	    </li>
          
            <li class="toctree-l1 current">
		
    <a class="current" href="./">API</a>
    <ul class="subnav">
            
    <li class="toctree-l2"><a href="#the-genomicvectors-types-and-methods">The GenomicVectors Types and Methods</a></li>
    
        <ul>
        
            <li><a class="toctree-l3" href="#index">Index</a></li>
        
            <li><a class="toctree-l3" href="#types">Types</a></li>
        
            <li><a class="toctree-l3" href="#genome-location-api">Genome Location API</a></li>
        
            <li><a class="toctree-l3" href="#accessing-position-info">Accessing position info</a></li>
        
            <li><a class="toctree-l3" href="#modifying-positions">Modifying positions</a></li>
        
            <li><a class="toctree-l3" href="#querying-positions">Querying positions</a></li>
        
        </ul>
    

    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../future/">The Future</a>
	    </li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">GenomicRanges.jl</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
    
    <li>API</li>
    <li class="wy-breadcrumbs-aside">
      
        <a href="https://github.com/phaverty/GenomicVectors.jl/edit/master/docs/api.md"
          class="icon icon-github"> Edit on GitHub</a>
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <p><a id='The-GenomicVectors-Types-and-Methods-1'></a></p>
<h1 id="the-genomicvectors-types-and-methods">The GenomicVectors Types and Methods</h1>
<p><a id='Index-1'></a></p>
<h2 id="index">Index</h2>
<ul>
<li><a href="./#GenomicVectors.GenomeInfo"><code>GenomicVectors.GenomeInfo</code></a></li>
<li><a href="./#GenomicVectors.GenomicPositions"><code>GenomicVectors.GenomicPositions</code></a></li>
<li><a href="./#GenomicVectors.GenomicRanges"><code>GenomicVectors.GenomicRanges</code></a></li>
<li><a href="./#GenomicVectors.chrpos"><code>GenomicVectors.chrpos</code></a></li>
<li><a href="./#GenomicVectors.genopos"><code>GenomicVectors.genopos</code></a></li>
<li><a href="./#GenomicVectors.nearest"><code>GenomicVectors.nearest</code></a></li>
<li><a href="./#GenomicVectors.overlap"><code>GenomicVectors.overlap</code></a></li>
<li><a href="./#GenomicVectors.overlapin"><code>GenomicVectors.overlapin</code></a></li>
<li><a href="./#RLEVectors.ends"><code>RLEVectors.ends</code></a></li>
<li><a href="./#RLEVectors.starts"><code>RLEVectors.starts</code></a></li>
<li><a href="./#RLEVectors.widths"><code>RLEVectors.widths</code></a></li>
</ul>
<p><a id='Types-1'></a></p>
<h2 id="types">Types</h2>
<p><a id='GenomicVectors.GenomeInfo' href='#GenomicVectors.GenomeInfo'>#</a>
<strong><code>GenomicVectors.GenomeInfo</code></strong> &mdash; <em>Type</em>.</p>
<p><strong>GenomeInfo Type</strong></p>
<p>A GenomeInfo holds information about a genome including its name, chromosome names,  chromosome lengths and chromosome offsets into a concatenated, linear genome (genopos). Indexing returns the genopos end of the indexed chromosome.</p>
<p><strong>Examples</strong></p>
<pre><code class="julia">chrinfo = GenomeInfo(&quot;hg19&quot;,[&quot;chr1&quot;,&quot;chr2&quot;,&quot;chrX&quot;],Int64[3e5,2e5,1e4])
genome(chrinfo)
chr_names(chrinfo)
chr_lengths(chrinfo)
chr_ends(chrinfo)
chr_offsets(chrinfo)
chrinfo[2] # 5e5
</code></pre>

<p><a target='_blank' href='https://github.com/phaverty/GenomicVectors.jl/tree/6b9e9213367db41f7779769234164ec95502fab6/src/GenomeInfo_type.jl#L2-L18' class='documenter-source'>source</a><br></p>
<p><a id='GenomicVectors.GenomicPositions' href='#GenomicVectors.GenomicPositions'>#</a>
<strong><code>GenomicVectors.GenomicPositions</code></strong> &mdash; <em>Type</em>.</p>
<pre><code>GenomicPositions(chrpos, chromosomes, genomeinfo)
GenomicPositions(genopos, genomeinfo)
</code></pre>

<p>Represents single-nucleotide positions in a genome.</p>
<p>This type uses its (immutable) <code>GenomeInfo</code> slot object to describe corresponding genome and     positions can be expressed relative to this concatenated, linearized genome or relative     to the chromosome containing a given position.</p>
<p>Sorting is by chromosome, as ordered by chrinfo,</p>
<p>By convention, all postions in a <code>GenomicPositions</code> are considered to be on the plus strand.</p>
<hr />
<p><strong>Examples</strong></p>
<pre><code class="julia">    genomeinfo = GenomeInfo(&quot;hg19&quot;,[&quot;chr1&quot;,&quot;chr2&quot;,&quot;chrX&quot;],Int64[3e5,2e5,1e4])
    chrs = [&quot;chr2&quot;,&quot;chr1&quot;,&quot;chr2&quot;,&quot;chrX&quot;]
    pos = Int64[3e4,4.2e3,1.9e5,1e4]
    gpos = genopos(pos,chrs,chrinfo)
    x = GenomicPositions(pos,chrs,genomeinfo)
    y = GenomicPositions(gpos,genomeinfo)
    same_genome(x, y)
    sort!(y)
    convert(DataFrame, y)
</code></pre>

<p><a target='_blank' href='https://github.com/phaverty/GenomicVectors.jl/tree/6b9e9213367db41f7779769234164ec95502fab6/src/GenomicPositions_type.jl#L5-L33' class='documenter-source'>source</a><br></p>
<p><a id='GenomicVectors.GenomicRanges' href='#GenomicVectors.GenomicRanges'>#</a>
<strong><code>GenomicVectors.GenomicRanges</code></strong> &mdash; <em>Type</em>.</p>
<p><strong><code>GenomicRanges</code></strong></p>
<p><code>GenomicRanges</code> represent closed ranges in a genome. This type uses its (immutable) <code>GenomeInfo</code> slot object to describe         corresponding genome and positions can be expressed relative to this concatenated, linearized genome or relative to the chromosome containing a given position.</p>
<p><strong>Examples</strong></p>
<pre><code>chrinfo = GenomeInfo(&quot;hg19&quot;,[&quot;chr1&quot;,&quot;chr2&quot;,&quot;chrX&quot;],Int64[3e5,2e5,1e4])
chrs = [&quot;chr1&quot;,&quot;chr2&quot;,&quot;chr2&quot;,&quot;chrX&quot;]
starts = [100, 200, 300, 400]
ends = [120, 240, 350, 455]
gr = GenomicRanges(chrs,starts,ends,chrinfo)
</code></pre>

<p><strong>Indexing</strong></p>
<p>Indexing a <code>GenomicRanges</code> with an array produces a new <code>GenomicRanges</code>. Indexing by a scalar produces a two-tuple of the start and end positions in genome location units.</p>
<p><a target='_blank' href='https://github.com/phaverty/GenomicVectors.jl/tree/6b9e9213367db41f7779769234164ec95502fab6/src/GenomicRanges_type.jl#L5-L20' class='documenter-source'>source</a><br></p>
<p><a id='Genome-Location-API-1'></a></p>
<h2 id="genome-location-api">Genome Location API</h2>
<p><code>GenomicVectors.jl</code> has ... All <code>AbstractGenomicVector</code>s implement the API for <code>GenomeInfo</code> for access to their genome descriptions.</p>
<p><a id='Accessing-position-info-1'></a></p>
<h2 id="accessing-position-info">Accessing position info</h2>
<pre><code>starts
widths
ends
genopos
chrpos
chromosomes
</code></pre>

<p><a id='Modifying-positions-1'></a></p>
<h2 id="modifying-positions">Modifying positions</h2>
<pre><code>slide
slide!
</code></pre>

<p><a id='Querying-positions-1'></a></p>
<h2 id="querying-positions">Querying positions</h2>
<p>As in Bioconductor, location query operations discriminate between exact and overlapping matches. In addition to exact versus overlapping coordinates, exact matching includes strand, while overlap matching does not. In <code>GenomcVectors.jl</code>, the standard set operations use exact matching and custom overlap functions are defined for <code>AbstractGenomicVector</code>.</p>
<p><a id='Overlap-function-1'></a></p>
<h3 id="overlap-function">Overlap function</h3>
<pre><code>overlap
overlapin
overlapindex
nearest
</code></pre>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../future/" class="btn btn-neutral float-right" title="The Future">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href="../TODO/" class="btn btn-neutral" title="TODO"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
	  
        </div>
      </div>

    </section>
    
  </div>

  <div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
          <a href="https://github.com/phaverty/GenomicVectors.jl" class="fa fa-github" style="float: left; color: #fcfcfc"> GitHub</a>
      
      
        <span><a href="../TODO/" style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../future/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script src="../js/theme.js"></script>
      <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
      <script src="../assets/mathjaxhelper.js"></script>

</body>
</html>
